clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java

```diff
@@ -173,14 +172,38 @@ public void dispose() throws SaslException {
             throw new SaslAuthenticationException(String.format(
                     "Authentication failed: Client requested an authorization id (%s) that is different from the token's principal name (%s)",
                     authorizationId, token.principalName()));
+
+        Map<String, String> validExtensions = processExtensions(token, extensions);
+
         tokenForNegotiatedProperty = token;
-        this.extensions = extensions;
+        this.extensions = new SaslExtensions(validExtensions);
         complete = true;
         if (log.isDebugEnabled())
             log.debug("Successfully authenticate User={}", token.principalName());
         return new byte[0];
     }
 
+    private Map<String, String> processExtensions(OAuthBearerToken token, SaslExtensions extensions) throws SaslException {
+        OAuthBearerExtensionsValidatorCallback extensionsCallback = new OAuthBearerExtensionsValidatorCallback(token, extensions);
+        try {
+            callbackHandler.handle(new Callback[] {extensionsCallback});
+        } catch (UnsupportedCallbackException e) {
+            // backwards compatibility - no extensions will be added
+        } catch (IOException e) {
+            handleInternalError(e);
+        }
+        if (!extensionsCallback.invalidExtensions().isEmpty()) {
+            String errorMessage = String.format("Authentication failed: %d extensions are invalid! They are:%n%s",
```

Comment:
%n adds a newline?  Not sure this is a good idea in a log message, best to make it a space, I think.

Reply:
Why not, though?
A space could be better but my reasoning was that a new line makes it more clear